src/vos/tests/csum_extent_tests.c

```diff
@@ -942,3 +942,78 @@ void csum_invalid_input_tests(void **state)
 	free(read_data_buf);
 	free(csum_buf_1);
 }
+
+/* Inject fault on the update and fetch.
+* The update will initialize checksum value to zero.
+* Fetch operation will return random values in checksum field.
+* Test will compare the update and fetch output for pass/fail
+*/
+void
+csum_fault_injection_multiple_extents_tests(void **state)
+{
+	int rc;
+	int i;
+	int update_extents;
+	int fetch_extents;
+	struct csum_test test;
+	/* Setup Test */
+	struct csum_test_params params;
+	uint32_t csums_count_total, csum_count_per_extent, csum_len;
+	/* extent counts for update and fetch. The extents will span the same
+	 * amount of data, the extents themselves will be of different lengths
+	 */
+	const int END = 0;
+	const int table[][2] = {
+	/*	update, fetch  */
+		{1,	1},
+		{1,	4},
+		{2,	4},
+		{4,	4},
+		{4,	1},
+		{4,	2},
+		{END,	END}
+	};
+
+
+	params.total_records = 1024 * 1024 * 64;
+	params.record_bytes = 1;
+	params.csum_bytes = 8; /* CRC64? */
+	params.csum_chunk_records = 1024 * 16; /* 16K */
+	params.use_rand_csum = true;
+
+	csum_test_setup(&test, state, &params);
+
+
+	for (i = 0; table[i][0] != END; i++) {
+		update_extents = table[i][0];
+		fetch_extents = table[i][1];
+
+		printf("Update Extents: %d, Fetch Extents: %d\n",
+		       update_extents, fetch_extents);
+
+		daos_fail_loc_set(DAOS_CHECKSUM_UPDATE_FAIL | DAOS_FAIL_ONCE);
+		rc = update(&test, update_extents, i);
+		if (!SUCCESS(rc))
+			fail_msg("Error updating extent with csum: %s\n",
+				 d_errstr(rc));
+
+		daos_fail_loc_set(DAOS_CHECKSUM_FETCH_FAIL | DAOS_FAIL_ONCE);
+		rc = fetch(&test, fetch_extents, &csum_count_per_extent,
+			   &csum_len, &csums_count_total, i);
+		if (!SUCCESS(rc))
+			fail_msg("Error fetching extent with csum: %s\n",
+				 d_errstr(rc));
+
+		/* Verify */
+		assert_int_equal(test.csum_bytes, csum_len);
+		assert_int_equal(get_csums_per_extent(&test, fetch_extents),
+				 csum_count_per_extent);
+		assert_int_equal(get_csum_total(&test, fetch_extents),
+				csums_count_total);
+		assert_memory_not_equal(test.update_csum_buf,
```

Comment:
I don't think this fully verifies the test here. It does verify the checksum is corrupted, but does it know if it happened on update or fetch? If the error inject on update doesn't actually work, the fetch error inject would hide that and vice versa. Or am I missing something?

Reply:
The problem here is all tests are running on the VOS. I believe when we have the client/server model fully functional, we should be able to create tests with update fault and see whether fetch gives different value. Presently, all the tests are running on VOS (even if we inject fault on update, fetch is going to get the bad data unless we can tell what fetch is supposed get as value). Maybe when we calculate actual checksum under server, we can enhance/add the test cases. My few cents.

Reply:
I'm not sure I grok the reasoning but it's Friday afternoon.